base/compiler/typelattice.jl

```diff
@@ -73,14 +73,18 @@ struct Conditional
     slot::Int
     thentype
     elsetype
-    function Conditional(slot::Int, @nospecialize(thentype), @nospecialize(elsetype))
+    # `isdefined` indicates this `Conditional` is from `@isdefined slot`, implying that
+    # the `undef` information of `slot` can be improved in the then branch.
+    # Since this is only beneficial for local inference, it is not translated into `InterConditional`.
+    isdefined::Bool
```

Comment:
I'm not sure this works conceptually. `isdefined`-ness is not a flow property and can be reset by NewvarNode

Reply:
Basically to optimize isdefined you need to do ssa construction

Reply:
I thought that is a data-flow property. My understanding was that the type/isdefined-ness information for a slot gets reset by `NewvarNode`, but then both of those get updated along with the abstract interpretation. E.g. within `smerge`, both `typ` and `undef` are being merged.

The purpose of this PR was to eliminate unnecessary `:throw_undef_if_not` during SSA construction, but I might be making a fundamental mistake.

Reply:
Turns out `Conditional` itself is just broken: https://github.com/JuliaLang/julia/issues/55548